network/parallelism.go

```diff
@@ -0,0 +1,263 @@
+	p.statshub.SendParralelismStats(types.ParralelismStats{
+		MinConnections:     p.cfg.MinConnections,
+		MaxConnections:     p.cfg.MaxConnections,
+		DesiredConnections: p.currentDesired,
+	})
+}
+
+func (p *parallelism) desiredLoops() {
+	p.mut.Lock()
+	defer p.mut.Unlock()
+
+	// Dont bother calculating if connections are the same value.
+	if p.cfg.MinConnections == p.cfg.MaxConnections {
+		level.Debug(p.l).Log("msg", "min and max loops are same no change", "desired", p.cfg.MinConnections)
+		p.changeParallelism(p.currentDesired)
+	}
+
+	// Loop over network errors and remove them if the ttl expired.
+	keepErrors := make([]time.Time, 0, len(p.networkErrors))
+	for _, err := range p.networkErrors {
+		if time.Since(err) <= p.cfg.ResetInterval {
+			keepErrors = append(keepErrors, err)
+		}
+	}
+	p.networkErrors = keepErrors
+
+	keepSuccesses := make([]time.Time, 0, len(p.networkSuccesses))
+	for _, err := range p.networkSuccesses {
+		if time.Since(err) <= p.cfg.ResetInterval {
+			keepSuccesses = append(keepSuccesses, err)
+		}
+	}
+	p.networkSuccesses = keepSuccesses
+	errorRate := p.networkErrorRate()
+	// If we have network errors then ramp down the number of loops.
+	if p.cfg.AllowedNetworkErrorPercent != 0.0 && errorRate >= p.cfg.AllowedNetworkErrorPercent {
+		// Need to keep the value between min and max.
+		if p.currentDesired-1 >= p.cfg.MinConnections {
+			level.Debug(p.l).Log("msg", "triggering lower desired due to network errors", "desired", p.currentDesired-1)
+			p.changeParallelism(p.currentDesired - 1)
+		}
+		return
+	}
+	// If we are drifting too much then ramp up the number of loops.
+	if p.timestampDriftSeconds > p.cfg.AllowedDriftSeconds {
+		// Need to keep the value between min and max.
+		if p.currentDesired+1 <= p.cfg.MaxConnections {
+			level.Debug(p.l).Log("msg", "increasing desired due to timestamp drift", "desired", p.currentDesired+1, "drift", p.timestampDriftSeconds)
+			p.changeParallelism(p.currentDesired + 1)
+		}
+		return
+	}
+
+	// Can we ramp down, if we have dropped below our minimum scale down.
+	if p.timestampDriftSeconds < p.cfg.MinimumScaleDownDriftSeconds {
+		// Need to keep the value between min and max.
+		if p.currentDesired-1 >= p.cfg.MinConnections {
+			level.Debug(p.l).Log("msg", "decreasing desired due to drift lowering", "desired", p.currentDesired-1, "drift", p.timestampDriftSeconds)
+			p.changeParallelism(p.currentDesired - 1)
+		}
+	}
+	level.Debug(p.l).Log("msg", "no changes needed", "desired", p.currentDesired)
+}
+
+func (p *parallelism) networkErrorRate() float64 {
+	// If nothing has happened assume success
+	if len(p.networkSuccesses) == 0 && len(p.networkErrors) == 0 {
+		return 0.0
+	}
+
+	if len(p.networkErrors) == 0 {
+		return 0.0
+	}
+
+	if len(p.networkSuccesses) == 0 {
+		return 1.0
+	}
+
+	errorRate := float64(len(p.networkErrors)) / float64(len(p.networkSuccesses)+len(p.networkErrors))
+	return errorRate
+}
+
+func (p *parallelism) changeParallelism(desired uint) {
```

Comment:
```suggestion
func (p *parallelism) changeParallelism(proposed uint) {
```

I find it hard to follow when we have `p.currentDesired` and `desired`. Maybe let's use 'proposed'?

Reply:
Sounds reasonable.